dart/dynamics/HierarchicalIK.cpp

```diff
@@ -0,0 +1,809 @@
+    assert(false);
+    return 0.0;
+  }
+
+  const IKHierarchy& hierarchy = hik->getIKHierarchy();
+
+  double cost = 0.0;
+  for(size_t i=0; i < hierarchy.size(); ++i)
+  {
+    const std::vector< std::shared_ptr<InverseKinematics> >& level =
+        hierarchy[i];
+
+    for(size_t j=0; j < level.size(); ++j)
+    {
+      const std::shared_ptr<InverseKinematics>& ik = level[j];
+
+      if(!ik->isActive())
+        continue;
+
+      const std::vector<size_t>& dofs = ik->getDofs();
+      Eigen::VectorXd q(dofs.size());
+      for(size_t k=0; k < dofs.size(); ++k)
+        q[k] = _x[dofs[k]];
+
+      InverseKinematics::ErrorMethod& method = ik->getErrorMethod();
+      const Eigen::Vector6d& error = method.evalError(q);
+
+      cost += error.dot(error);
+    }
+  }
+
+  return std::sqrt(cost);
+}
+
+//==============================================================================
+void HierarchicalIK::Constraint::evalGradient(
+    const Eigen::VectorXd& _x, Eigen::Map<Eigen::VectorXd> _grad)
+{
+  const std::shared_ptr<HierarchicalIK>& hik = mIK.lock();
+
+  const IKHierarchy& hierarchy = hik->getIKHierarchy();
+  const SkeletonPtr& skel = hik->getSkeleton();
+  const size_t nDofs = skel->getNumDofs();
+  const std::vector<Eigen::MatrixXd>& nullspaces = hik->computeNullSpaces();
+
+  _grad.setZero();
+  for(size_t i=0; i < hierarchy.size(); ++i)
+  {
+    const std::vector< std::shared_ptr<InverseKinematics> >& level =
+        hierarchy[i];
+
+    mLevelGradCache.setZero(nDofs);
+    for(size_t j=0; j < level.size(); ++j)
+    {
+      const std::shared_ptr<InverseKinematics>& ik = level[j];
+
+      if(!ik->isActive())
+        continue;
+
+      // Grab only the dependent coordinates from q
+      const std::vector<size_t>& dofs = ik->getDofs();
+      Eigen::VectorXd q(dofs.size());
+      for(size_t k=0; k < dofs.size(); ++k)
+        q[k] = _x[dofs[k]];
+
+      // Compute the gradient of this specific error term
+      mTempGradCache.setZero(dofs.size());
+      Eigen::Map<Eigen::VectorXd> gradMap(mTempGradCache.data(),
+                                          mTempGradCache.size());
+
+      InverseKinematics::GradientMethod& method = ik->getGradientMethod();
+      method.evalGradient(q, gradMap);
+
+      // Add the components of this gradient into the gradient of this level
+      for(size_t k=0; k < dofs.size(); ++k)
+        mLevelGradCache[dofs[k]] += mTempGradCache[k];
+    }
+
+    // Project this level's gradient through the null spaces of the levels with
+    // higher precedence, then add it to the overall gradient
+    if(i > 0)
+      _grad += nullspaces[i-1] * mLevelGradCache;
+    else
+      _grad += mLevelGradCache;
+  }
+}
+
+//==============================================================================
+HierarchicalIK::HierarchicalIK(const SkeletonPtr& _skeleton)
+  : mSkeleton(_skeleton)
+{
+  // initialize MUST be called from the constructors of any directly inheriting
+  // classes
```

Comment:
The comment isn't quite right. ``initialize`` must be called immediately after the construction of a HierarchicalIK instance. This is why the derived classes have ``create()`` functions and protected constructors.